### PR TITLE
Add customizable prefix to oban-emitted telemetry events

### LINKED ISSUE #545: Support a custom prefix for telemetry events
### Is your feature request related to a problem? Please describe.

It's difficult to differentiate between events from multiple oban instances when using systems like Phoenix Live Dashboard. All events share a common path, e.g. `[:oban, :job, :stop]`.

### Describe the Solution You'd Like

Support a customizable event prefix that is included in all telemetry events. The default will be `[:oban]`, to match the current functionality.

### Describe Alternatives You've Considered

Matching on the `conf.name` metadata. This works for ad-hoc implementations, but not for path-based measurements like Live Dashboard uses.

### PR DESCRIPTION
**Pull request by v0idpwn**
Closes #545.

### CODE REVIEW

File: lib/oban/telemetry.ex [hunk 1]
```diff
@@ -305,16 +305,20 @@ defmodule Oban.Telemetry do
...
       :ok = Oban.Telemetry.attach_default_logger(:debug)
+
+  Attach a logger with a specific oban prefix:
+
+      :ok = Oban.Telemetry.attach_default_logger(:info, [:my_oban])
   """
   @doc since: "0.4.0"
   @spec attach_default_logger(Logger.level()) :: :ok | {:error, :already_exists}
-  def attach_default_logger(level \\ :info) do
+  def attach_default_logger(level \\ :info, telemetry_prefix \\ [:oban]) do
```
Comment (sorentwo) [inline lib/oban/telemetry.ex:315]: Using two defaults makes it impossible to set the prefix without setting the level. What if we turn this into a keyword list instead?

```elixir
  def attach_default_logger(level) when is_atom(level) do
    attach_default_logger(level: level)
  end

  def attach_default_logger(opts) when is_list(opts) do
    # ...
```

File: lib/oban/telemetry.ex [hunk 2]
```diff
@@ -370,17 +374,19 @@ defmodule Oban.Telemetry do
...
-    |> Map.merge(converted_measurements(measure))
-    |> log_message("job:#{event}", level)
-  end
-
-  def handle_event([:oban, :circuit, event], _measure, meta, level) do
-    meta
-    |> Map.take([:message, :name])
-    |> log_message("circuit:#{event}", level)
+  def handle_event(event_name, measure, meta, level) do
+    case Enum.slice(event_name, -2..-1) do
```
Comment (sorentwo) [inline lib/oban/telemetry.ex:378]: I'm concerned about the performance of `Enum.slice` when processing a _lot_ of events. It's a bit manual, but we could specify clauses for prefixes with 1-3 values and put a limit in the `config` validation. Something like this

```elixir
[_, :job, event] -> # ...
[_, _, :job, event] -> # ...
[_, _, _, :job, event] -> # ...
```

That may be overkill, and I'm curious to see a benchmark that proves `Enum.slice` is good enough. WDYT?

File: test/oban/config_test.exs [hunk 3]
```diff
@@ -81,6 +81,13 @@ defmodule Oban.ConfigTest do
       assert_valid(shutdown_grace_period: 10)
     end
 
+    test ":telemetry_prefix is validated as a list of atoms" do
+      refute_valid(telemetry_prefix: :oban)
+      refute_valid(telemetry_prefix: ["oban"])
+
+      assert_valid(telemetry_prefix: [:oban])
```
Comment (sorentwo) [inline test/oban/config_test.exs:88]:
```suggestion
      assert_valid(telemetry_prefix: [:oban])
      assert_valid(telemetry_prefix: [:my_app, :oban])
```

**Review by sorentwo — COMMENTED** [review]
Nicely done! You've covered just about everything. I have one suggestion for telemetry and an open performance question about `Enum.slice`. Otherwise, this is good to go 👍

File: lib/oban/telemetry.ex (hunk 1 above)
Comment (v0idpwn) [reply lib/oban/telemetry.ex:315]: That's a good idea! I was afraid of breaking compatibility, but your default solves it :)

File: lib/oban/telemetry.ex (hunk 2 above)
Comment (v0idpwn) [reply lib/oban/telemetry.ex:378]: I was not sure about which was the best solution there. I'll try to come up with a benchmark

Comment (v0idpwn) [reply lib/oban/telemetry.ex]: Pattern matching is around 5-6x faster than relying on Enum.slice. 

I had two options: 
- defining a single clause and using a case
- defining 6 clauses

I went with the second one.

**Comment by v0idpwn** [comment]
I think I addressed all your comments. Let me know if there's something else you'd like to see changed :).

Probably more docs need to be updated as well. I don't know what should be done with the countless references to `[:oban, some, event]` telemetry events there.

File: test/oban/config_test.exs [hunk 4]
```diff
@@ -81,6 +81,15 @@ defmodule Oban.ConfigTest do
...
     end
 
+    test ":telemetry_prefix is validated as a list of atoms" do
+      refute_valid(telemetry_prefix: :oban)
+      refute_valid(telemetry_prefix: ["oban"])
+      refute_valid(telemetry_prefix: [:more, :than, :three, :elements])
+
+      assert_valid(telemetry_prefix: [:oban])
+      assert_valid(telemetry_prefix: [:my_app, :oban])
+    end
```
Comment (v0idpwn) [inline test/oban/config_test.exs:91]: Missed a case: `[]`

File: lib/oban/config.ex [hunk 5]
```diff
@@ -205,9 +205,9 @@ defmodule Oban.Config do
   end
 
   defp validate_opt!({:telemetry_prefix, prefix}) do
-    unless is_list(prefix) and Enum.all?(prefix, &is_atom/1) and length(prefix) <= 3 do
+    unless is_list(prefix) and Enum.all?(prefix, &is_atom/1) and length(prefix) in 1..3 do
       raise ArgumentError,
-            "expected :telemetry_prefix to be a list of atoms with size <= 3, got: #{inspect(prefix)}"
+            "expected :telemetry_prefix to be a non-empty list of atoms with less than 3 items, got: #{inspect(prefix)}"
```
Comment (sorentwo) [inline lib/oban/config.ex:210]:
```suggestion
            "expected :telemetry_prefix to be a non-empty list of atoms with up to 3 items, got: #{inspect(prefix)}"
```

File: lib/oban/telemetry.ex [hunk 6]
```diff
@@ -302,19 +302,28 @@ defmodule Oban.Telemetry do
...
-  Attach a logger at the `:debug` level:
+  Attach a logger with a specific oban prefix and level:
 
-      :ok = Oban.Telemetry.attach_default_logger(:debug)
+      :ok = Oban.Telemetry.attach_default_logger(telemetry_prefix: [:my_oban], logger_level: :debug)
   """
   @doc since: "0.4.0"
   @spec attach_default_logger(Logger.level()) :: :ok | {:error, :already_exists}
-  def attach_default_logger(level \\ :info) do
+  def attach_default_logger(level_or_opts \\ :info)
```
Comment (sorentwo) [inline lib/oban/telemetry.ex:311]:
```suggestion
  def attach_default_logger(level_or_opts \\ [logger_level: :info])
```

**Review by sorentwo — APPROVED** [review]

File: lib/oban/telemetry.ex [hunk 7]
```diff
@@ -302,19 +302,28 @@ defmodule Oban.Telemetry do
...
-  Attach a logger at the `:debug` level:
+  Attach a logger with a specific oban prefix and level:
 
-      :ok = Oban.Telemetry.attach_default_logger(:debug)
+      :ok = Oban.Telemetry.attach_default_logger(telemetry_prefix: [:my_oban], logger_level: :debug)
   """
   @doc since: "0.4.0"
   @spec attach_default_logger(Logger.level()) :: :ok | {:error, :already_exists}
-  def attach_default_logger(level \\ :info) do
+  def attach_default_logger(level_or_opts \\ [logger_level: :info])
```
Comment (sorentwo) [inline lib/oban/telemetry.ex:311]: We'll need a different typespec here:

```elixir
  @type attach_option :: {:logger_level, Logger.level()} | {:telemetry_prefix, [atom()]}
```

```elixir
  @spec attach_default_logger(Logger.level() | [attach_option()]) :: :ok | {:error, :already_exists}
```

**Comment by sorentwo** [comment]
Wonderful, thank you! 💛

**Pull request merged**